File: frontend/src/services/flightTrackingService.ts
import { apiClient } from './api';
import { ApiResponse, TrackedFlight } from '../types/api';

export const flightTrackingService = {
  /**
   * Track a flight to receive live status updates.
   */
  async trackFlight(flightId: string): Promise<TrackedFlight> {
    const response = await apiClient.post<ApiResponse<TrackedFlight>>(`/v1/flights/${flightId}/track`);
    return response.data.data;
  },

  /**
   * Stop tracking a flight.
   */
  async untrackFlight(flightId: string): Promise<void> {
    await apiClient.delete<ApiResponse<void>>(`/v1/flights/${flightId}/track`);
  },

  /**
   * Get all tracked flights for current user with live status.
   */
  async getTrackedFlights(): Promise<TrackedFlight[]> {
    const response = await apiClient.get<ApiResponse<TrackedFlight[]>>('/v1/flights/tracked');
    return response.data.data || [];
  },

  /**
   * Check if a flight is currently tracked by user.
   */
  async isTracking(flightId: string): Promise<boolean> {
    try {
      const response = await apiClient.get<ApiResponse<boolean>>(`/v1/flights/${flightId}/track/status`);
      return Boolean(response.data.data);
    } catch {
      return false;
    }
  },
};

File: backend/src/main/java/com/smarttravel/modules/flight/tracking/service/FlightTrackingService.java
package com.smarttravel.modules.flight.tracking.service;

import com.smarttravel.modules.flight.tracking.dto.TrackedFlightResponse;

import java.util.List;

/**
 * Service for managing user flight tracking subscriptions.
 */
public interface FlightTrackingService {

    /**
     * Track a flight for the current user. Idempotent — tracking an already tracked
     * flight returns the existing record.
     */
    TrackedFlightResponse trackFlight(String flightId, String userId);

    /**
     * Stop tracking a flight.
     */
    void untrackFlight(String flightId, String userId);

    /**
     * Get all currently tracked flights for a user, populated with live flight status.
     */
    List<TrackedFlightResponse> getTrackedFlights(String userId);

    /**
     * Check whether a user is tracking a specific flight.
     */
    boolean isTracking(String flightId, String userId);
}
